lib/srv/db/mongodb/connect.go

```diff
@@ -153,6 +161,43 @@ func (e *Engine) getConnectionOptions(ctx context.Context, sessionCtx *common.Se
 	}, nil
 }
 
+func (e *Engine) getAuthenticator(ctx context.Context, sessionCtx *common.Session) (authenticator auth.Authenticator, err error) {
+	// Currently, the MongoDB Atlas IAM Authentication doesn't work with IAM
+	// users. Here we provide a better error message to the users.
+	if sessionCtx.Database.GetType() == types.DatabaseTypeMongoAtlas && awsutils.IsUserARN(sessionCtx.DatabaseUser) {
+		return nil, trace.BadParameter("MongoDB Atlas AWS IAM Authentication with IAM users is not supported.")
+	}
+
+	switch {
+	case sessionCtx.Database.GetType() == types.DatabaseTypeMongoAtlas && awsutils.IsRoleARN(sessionCtx.DatabaseUser):
+		e.Log.Debug("Authenticating to database using AWS IAM authentication")
+		username, password, sessToken, authErr := e.Auth.GetAtlasIAMToken(ctx, sessionCtx)
+		if err != nil {
+			return nil, trace.Wrap(authErr)
+		}
+
+		authenticator, err = auth.CreateAuthenticator(auth.MongoDBAWS, &auth.Cred{
+			Source:   awsIAMSource,
+			Username: username,
+			Password: password,
+			Props: map[string]string{
+				awsSecretTokenKey: sessToken,
+			},
+		})
+	default:
+		e.Log.Debug("Authenticating to database using certificates")
+		authenticator, err = auth.CreateAuthenticator(auth.MongoDBX509, &auth.Cred{
+			// MongoDB uses full certificate Subject field as a username.
+			Username: "CN=" + sessionCtx.DatabaseUser,
+		})
+	}
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
```

Comment:
nit: 
I know that this helps for duplication but due to scope locality and readability i prefer sequence flow lie: function call  followed by error handling but leaving it up to you.